[![Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/sdoro/notebooks/blob/master/bigData.ipynb)   [![Binder](http://mybinder.org/badge.svg)](http://mybinder.org/v2/gh/sdoro/notebooks/master?filepath=bigData.ipynb)


# Prefissi della unità di misura (decimali)

I [Prefissi nel Sistema Internazionale (di misura)](https://it.wikipedia.org/wiki/Prefissi_del_Sistema_internazionale_di_unit%C3%A0_di_misura) hanno nomi molto poco comuni. Completano estendendo i nomi i [bilioni](https://it.wikipedia.org/wiki/Bilione) e le voci correlate. Un altro riferimento è la categoria dei [numeri grandi](https://it.wikipedia.org/wiki/Categoria:Numeri_grandi).




### Riassunto

| Numero decimale | Scala lunga | Prefisso italiano | $10^n$ |
|  ---------------------------  | ------------------ | -------------------------- | ------------ | 
| 1 000 000 000 000 000 000 000 000 000 000 000 | Quintiliardo | -  | $10^{33}$ |
| 1 000 000 000 000 000 000 000 000 000 000 | Quintilione | - | $10^{30}$ |
| 1 000 000 000 000 000 000 000 000 000 | Quadriliardo | - | $10^{27}$ |
| 1 000 000 000 000 000 000 000 000 | Quadrilione | yotta | $10^{24}$ |
| 1 000 000 000 000 000 000 000 | Triliardo | zetta | $10^{21}$ |
| 1 000 000 000 000 000 000 | Trilione | exa | $10^{18}$ |
| 1 000 000 000 000 000 | Biliardo | peta | $10^{15}$ |
| 1 000 000 000 000 | Bilione | tera | $10^{12}$ |
| 1 000 000 000 | Miliardo | giga | $10^{9}$ |
| 1 000 000 | Milione | mega | $10^{6}$ |
| 1 000 | Mille | chilo | $10^{3}$ |









La quantità di dati che produciamo ogni giorno è davvero sbalorditiva. Ci sono 2,5 quintilioni di byte di dati creati ogni giorno al nostro ritmo attuale, ma tale ritmo sta solo accelerando con la crescita dell'Internet of Things (IoT). Solo negli ultimi due anni è stato generato il 90% dei dati nel mondo. Quanto vale in numeri ?

In [1]:
n = 2.5 * 10**30
print("Il valore di 2.5 quintilioni al giorno è: {:20,.0f}".format(n))
TRILIARDI = 10**21
n = 2.5 * 10**30 / TRILIARDI
print("Il valore di 2.5 quintilioni al giorno in triliardi è: {:13,.0f}\n".format(n))
n = 2.5 * 10**30 / 3600
print("Il valore di 2.5 quintilioni al secondo è: {:20,.0f}".format(n))
QUADRILIONI = 10**24
n = 2.5 * 10**30 / 3600 / QUADRILIONI
print("Il valore di 2.5 quintilioni al secondo è: {:4,.2f} quadrilioni".format(n))

Il valore di 2.5 quintilioni al giorno è: 2,499,999,999,999,999,908,974,073,741,312
Il valore di 2.5 quintilioni al giorno in triliardi è: 2,500,000,000

Il valore di 2.5 quintilioni al secondo è: 694,444,444,444,444,354,410,446,848
Il valore di 2.5 quintilioni al secondo è: 694.44 quadrilioni


In media, Google elabora più di 40.000 ricerche OGNI secondo (3,5 miliardi di ricerche al giorno)!

In [2]:
n = 40000 * 3600 * 24
print("Il numero di ricerche elaborate da Google al giorno è: {:10,.0f}".format(n))
GIGA = 10**9
n = 40000 * 3600 * 24 / GIGA
print("Il numero di ricerche elaborate da Google al giorno è: {:3,.2f} Giga\n".format(n))
n = 40000 * 3600 *24 * 365
print("Il numero di ricerche elaborate da Google in un anno è: {:10,.0f}".format(n))
TERA = 10**12
n = 40000 * 3600 *24 * 365 / TERA
print("Il numero di ricerche elaborate da Google in un anno è: {:3,.2f} Tera".format(n))

Il numero di ricerche elaborate da Google al giorno è: 3,456,000,000
Il numero di ricerche elaborate da Google al giorno è: 3.46 Giga

Il numero di ricerche elaborate da Google in un anno è: 1,261,440,000,000
Il numero di ricerche elaborate da Google in un anno è: 1.26 Tera


Dall'articolo [How Much Data Do We Create Every Day?](https://goo.gl/TpLFrk) si possono fare esercizi di calcolo interessanti e curiosi. I precedenti due calcoli ne sono un esempio.
Si possono costruire anche esempi di calcoli del tipo: "dopo una settimana ...", oppure "se la lunghezza fosse ..." o ancora "se la velocità fosse ..."

# Prefissi delle unità di misura (binarie)

I [Prefissi per multipli binari](https://it.wikipedia.org/wiki/Prefissi_per_multipli_binari).

[The Zettabyte Era: Trends and Analysis](https://www.cisco.com/c/en/us/solutions/collateral/service-provider/visual-networking-index-vni/vni-hyperconnectivity-wp.html) è un bel articolo (Cisco 2017) dove ci sono molti numeri esageratamente GRANDI.

### Riassunto

| Valore | Fattore | Simbolo | Nome esteso | Equivalente | Fattore | Errore |
| ---------- | -----------  | ------------- | --------------------- | ------------------- | -----------  | --------- |
| 1 024	| $2^{10}$ |	Ki	| kilobinary	| ≈	chilo	| $10^3$ |	+2,4%
| 1 048 576	| $2^{20}$ |	Mi	| megabinary	| ≈	mega	| $10^6$ |	+4,9%
| 1 073 741 824	| $2^{30}$ |	Gi	| gigabinary	| ≈	giga	| $10^9$	|+7,4%
| 1 099 511 627 776	| $2^{40}$ |	Ti	| terabinary	| ≈	tera	| $10^{12}$ |	+10,0%
| 1 125 899 906 842 624	| $2^{50}$ 	| Pi	| petabinary	| ≈	peta	| $10^{15}$	| +12,6%
| 1 152 921 504 606 846 976	| $2^{60}$ |	Ei	| exabinary	| ≈	exa	| $10^{18}$ |	+15,3%
| 1 180 591 620 717 411 303 424	| $2^{70}$ |	Zi	| zettabinary	| ≈	zetta	| $10^{21}$ | 	+18,1%
| 1 208 925 819 614 629 174 706 176	| $2^{80}$ |	Yi	| yottabinary	| ≈	yotta	| $10^{24}$ |	+20,9%


### Calcolo d'esempio reale

In [3]:
# the total IP traffic in 2016 was around 1.2ZB (Cisco 2017).
n = 1.2 * 2**70
print("Il traffico IP nel 2016 fu di circa: {:25,.0f} bytes (fonte Cisco 2017).".format(n))

Il traffico IP nel 2016 fu di circa: 1,416,709,944,860,893,511,680 bytes (fonte Cisco 2017).


# Ordini di grandezza

Un elenco di [Ordini di grandezza](https://it.wikipedia.org/wiki/Ordini_di_grandezza_(numeri&#41;) che offre spunti per esercizi.

### Astronomia
La galassia Andromeda che può essere individuata nel cielo notturno, in realtà si trova a una distanza di 2,54 milioni di anni luce. Un [anno luce](https://it.wikipedia.org/wiki/Anno_luce) è la distanza che la luce, che si muove a 300.000 km al secondo, percorre in un anno: approssimativamente 9,46 bilioni di km. Questo significa che la galassia Andromeda che vediamo oggi è quella di 2,54 milioni di anni fa! La luce che raggiunge i nostri occhi in questo momento, è stata emessa molto prima della comparsa dell'Homo Sapiens. Quando dialoghiamo con le stelle, in realtà ci colleghiamo con l'infinito passato.

In [4]:
d = 2540000
print("Distanza dalla Terra della galassia Andromeda: {0:9,.0f} anni luce ({0:.2e}).".format(d))

v = 300000
print("Velocità della luce: {0:7,.0f} km al secondo ({0:.0e}).".format(v))

ns_anno = 24*3600*365
print("Numero di secondi in un anno: {0:10,.0f} ({0:.0e}).".format(ns_anno))

dluce_anno = v * ns_anno
print("Distanza percorsa dalla luce in un anno: {0:10,.0f} in km (circa 9,4 bilioni di km) ({0:.1e}).".format(dluce_anno))

dkm = v * ns_anno * d
print("Distanza dalla Terra della galassia Andromeda: {0:10,.0f} in km (circa 24 triliardi di km) ({0:.1e}).".format(dkm))

Distanza dalla Terra della galassia Andromeda: 2,540,000 anni luce (2.54e+06).
Velocità della luce: 300,000 km al secondo (3e+05).
Numero di secondi in un anno: 31,536,000 (3e+07).
Distanza percorsa dalla luce in un anno: 9,460,800,000,000 in km (circa 9,4 bilioni di km) (9.5e+12).
Distanza dalla Terra della galassia Andromeda: 24,030,432,000,000,000,000 in km (circa 24 triliardi di km) (2.4e+19).


### Tecnologia di produzione delle CPU

All’interno del processore sono contenuti microscopici componenti, disposti su una griglia di quadrati che agiscono come switch On e Off, permettendo e ostacolando il passaggio della corrente elettrica dando luogo ad una sequenza di segnali elettrici codificati nel sistema binario in 0 e 1.

La dimensione media del gate di ogni singolo transistor si misura, per l’appunto, in [nanometri](https://it.wikipedia.org/wiki/Nanometro). Un nanometro non è altro che la miliardesima parte di un metro. Per capire quanto piccolo un nanometro sia, prendete come riferimento lo spessore medio di un capello umano: 80000 nm.

Dunque a parità di superficie, minore sarà la distanza tra i componenti del microprocessore maggiore sarà il loro numero. Ecco spiegato il motivo per cui processi produttivi più piccoli tendono a realizzare processori con una velocità di calcolo superiore.

### Calcolo die per wafer

Il [die](https://it.wikipedia.org/wiki/Die_(elettronica&#42;) è il sottile strato di materiale semiconduttore sul quale è realizzato il circuito elettronico. Nel caso di una CPU di tipo Xeon E7440 [processo a 45 nm ]è all'incirca quadrato: 22 x 22 mm; in un i7 8700K [processo a 22nm] è circa 12 x 12 mm. Il [numero di transistori](https://en.wikipedia.org/wiki/Transistor_count) è circa 2000 milioni.
Il [wafer](https://it.wikipedia.org/wiki/Wafer_(elettronica&#42;) è una sottile fetta di materiale semiconduttore di solito dal raggio tra 200 e 300 cm e su cui si stampano il numero massimo di die separati da una zona "di taglio" utile quando si separano i die.

### Esercizio
Dato un wafer di diametro 300 mm e un die di ampiezza 12 mm e un'area di taglio di 1mm, quanti die, al minimo, si possono produrre ?
Considerando l'intero wafer, quanti transistori si producono ?

In [5]:
import math
lato_quadrato_inscritto = 300 * math.sqrt(2)

die = 12
taglio = 1
numero_die_minimo = (int) (lato_quadrato_inscritto / (die + taglio))
numero_die_minimo *= numero_die_minimo
print("Il numero di die producibili per wafer (di 300 mm) è maggiore di: ", numero_die_minimo)

nt = 2000 * 10**6
ntt = numero_die_minimo * nt
print("Numero di transistori per wafer (di 300 mm) è maggiore di: {0:10,.0f} (circa 2 biliardi) ({0:.1e}).".format(ntt))

Il numero di die producibili per wafer (di 300 mm) è maggiore di:  1024
Numero di transistori per wafer (di 300 mm) è maggiore di: 2,048,000,000,000 (circa 2 biliardi) (2.0e+12).
